### PR TITLE
add interface for warning, error and report

### PR DESCRIPTION
**Pull request by mamueluth**
Add report interfaces for emergency stop, error and warning:
* The available report interfaces are: EMERGENCY_STOP_SIGNAL, ERROR_SIGNAL, ERROR_SIGNAL_MESSAGE,WARNING_SIGNAL and WARNING_SIGNAL_MESSAGE

The component interface (actuator_inteface, sensor_intefrace, system_inteface) creates and stores them locally. The interfaces are named like <hardware_name>/<report_interface_type>. E.g. if hardware is called rrbot -> interface for WARNING_SIGNAL is called: rrbot/WARNING_SIGNAL. Later with variant support the interfaces will hold following datatypes:
* Emergency Stop (bool)
* Error Code (vector<uint8>(32))
* Warning Code (vector<int8>(32))
* Error Message (vector<string(128)>(32)
* Warning Message (vector<string(128)>(32)

Where the <error/warn>_message_interfaces_ holds the _error/warn message_ at position X for the <error/warn>_interface's_ error signal at position X .

### CODE REVIEW

File: hardware_interface/include/hardware_interface/actuator_interface.hpp [hunk 1]
```diff
@@ -310,6 +351,33 @@ class ActuatorInterface : public rclcpp_lifecycle::node_interfaces::LifecycleNod
...
   }
 
+  void set_emergency_stop(const double & emergency_stop)
+  {
+    emergency_stop_->emergency_stop(emergency_stop);
+  }
+
+  double get_emergency_stop() const { return emergency_stop_->emergency_stop(); }
+
+  void set_warning_code(const std::string & warning_signal, const int8_t & warning_code)
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/actuator_interface.hpp]: This should be `uint8_t`.

File: hardware_interface/include/hardware_interface/handle.hpp [hunk 2]
```diff
@@ -27,7 +27,7 @@
 namespace hardware_interface
 {
 
-typedef std::variant<double> HANDLE_DATATYPE;
+typedef std::variant<bool, double, int8_t, uint8_t> HANDLE_DATATYPE;
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/handle.hpp]: If we already use variants, then this should use vectors of `uint8_t` and not individual values.

File: hardware_interface/include/hardware_interface/handle.hpp [hunk 3]
```diff
@@ -132,6 +132,39 @@ class StateInterface : public Handle
...
+  uint8_t error_code() const
+  {
+    const auto error_code = std::get_if<uint8_t>(&value_);
+    // This means the value does not store the expected datatype. How should we handle this
+    // properly?
+    THROW_ON_NOT_NULLPTR(error_code);
+    return *error_code;
+  }
+
+  void error_code(const double & error_code) { value_ = error_code; }
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/handle.hpp]: we have talked about this. This is not a feature of the interface but hardware.

File: hardware_interface/include/hardware_interface/loaned_state_interface.hpp [hunk 4]
```diff
@@ -65,6 +65,12 @@ class LoanedStateInterface
 
   double get_value() const { return state_interface_.get_value(); }
 
+  bool has_emergency_stop() const { return state_interface_.emergency_stop(); }
+
+  int8_t get_warning_code() const { return state_interface_.warning_code(); }
+
+  uint8_t get_error_code() const { return state_interface_.error_code(); }
+
```
Comment (destogl) [inline hardware_interface/include/hardware_interface/loaned_state_interface.hpp]:
```suggestion
```

**Review by destogl — CHANGES_REQUESTED** [review]
You are proposing to go with variants immediately. This is OK, but then we should do it properly. Variant support should be anyway done separately. 

That is the reason why I would not go with variants at first, but map everything to doubles.

Also, I am missing tests here to see how this actually works.

File: hardware_interface/include/hardware_interface/types/hardware_interface_error_signals.hpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+#define HARDWARE_INTERFACE__TYPES__HARDWARE_INTERFACE_ERROR_SIGNALS_HPP_
+
+#include <cstdint>
+#include <vector>
+
+namespace hardware_interface
+{
+constexpr char ERROR_SIGNAL_INTERFACE_NAME[] = "ERROR_SIGNAL";
+// Available error signal names
+enum class error_signal : std::uint8_t
```
Comment (mamueluth) [inline hardware_interface/include/hardware_interface/types/hardware_interface_error_signals.hpp:29]: The idea is for later if we support variant to set the vector at position hardware_interface::error_signal::ERROR_SIGNAL_0 to the error code

**Pull request closed without merging**